engine/src/main/java/org/pentaho/di/trans/steps/dimensionlookup/DimensionLookup.java

```diff
@@ -0,0 +1,1762 @@
+        }
+        data.preloadIndexes.add( index );
+      }
+
+      // This is all for now...
+    } catch ( Exception e ) {
+      throw new KettleException( "Error encountered during cache pre-load", e );
+    }
+  }
+
+  private synchronized Object[] lookupValues( RowMetaInterface rowMeta, Object[] row ) throws KettleException {
+    Object[] outputRow = new Object[ data.outputRowMeta.size() ];
+
+    RowMetaInterface lookupRowMeta;
+    Object[] lookupRow;
+
+    Object[] returnRow = null;
+
+    Long technicalKey;
+    Long valueVersion;
+    Date valueDate = null;
+    Date valueDateFrom = null;
+    Date valueDateTo = null;
+
+    // Determine the lookup date ("now") if we have a field that carries said
+    // date.
+    // If not, the system date is taken.
+    //
+    valueDate = determineDimensionUpdatedDate( row );
+
+    if ( !meta.isUpdate() && meta.isPreloadingCache() ) {
+      // Obtain a result row from the pre-load cache...
+      //
+      // Create a row to compare with
+      //
+      RowMetaInterface preloadRowMeta = data.preloadCache.getRowMeta();
+
+      // In this case it's all the same. (simple)
+      //
+      data.returnRowMeta = data.preloadCache.getRowMeta();
+      lookupRowMeta = preloadRowMeta;
+      lookupRow = new Object[ preloadRowMeta.size() ];
+
+      // Assemble the lookup row, convert data if needed...
+      //
+      for ( int i = 0; i < data.preloadIndexes.size(); i++ ) {
+        int from = data.preloadIndexes.get( i ); // Input row index
+        int to = data.preloadCache.getKeyIndexes()[ i ]; // Lookup row index
+
+        // From data type...
+        //
+        ValueMetaInterface fromValueMeta = rowMeta.getValueMeta( from );
+
+        // to date type...
+        //
+        ValueMetaInterface toValueMeta = data.preloadCache.getRowMeta().getValueMeta( to );
+
+        // From value:
+        //
+        Object fromData = row[ from ];
+
+        // To value:
+        //
+        Object toData = toValueMeta.convertData( fromValueMeta, fromData );
+
+        // Set the key in the row...
+        //
+        lookupRow[ to ] = toData;
+      }
+
+      // Also set the lookup date on the "end of date range" (toDate) position
+      //
+      lookupRow[ data.preloadFromDateIndex ] = valueDate;
+
+      // Look up the row in the pre-load cache...
+      //
+      int index = data.preloadCache.lookupRow( lookupRow );
+      if ( index >= 0 ) {
+        returnRow = data.preloadCache.getRow( index );
+      } else {
+        returnRow = null; // Nothing found!
+      }
+
+    } else {
+      lookupRow = new Object[ data.lookupRowMeta.size() ];
+      lookupRowMeta = data.lookupRowMeta;
+
+      // Construct the lookup row...
+      //
+      for ( int i = 0; i < meta.getKeyStream().length; i++ ) {
+        try {
+          lookupRow[ i ] = row[ data.keynrs[ i ] ];
+        } catch ( Exception e ) { // TODO : remove exception??
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: While wrapping the caught exception into a custom one, information about the caught exception is being lost, including information about the stack trace of the exception.

**Fix**: If the caught exception object does not contain sensitive information, consider passing it as the &#34;rootCause&#34; or inner exception parameter to the constructor of the new exception before throwing the new exception. (Note that not all exception constructors support inner exceptions. Use a wrapper exception that supports inner exceptions.) 
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)